src/vision/status_quo/barbara_builds_an_async_executor.md

```diff
@@ -0,0 +1,161 @@
+
+```rust
+pub struct Task {
+    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
+    scheduler: SchedulerSender,
+    priority: u8,
+}
+
+unsafe impl Sync for Task {}
+
+impl Wake for Task {
+    fn wake(self: Arc<Self>) {
+        self.scheduler.send(self.clone());
+    }
+}
+
+pub fn poll_task(task: Arc<Task>) {
+    let waker = Waker::from(task.clone());
+    let mut cx = Context::from_waker(&waker);
+    task.future.as_mut().poll(&mut cx);
+//  ^^^^^^^^^^^ cannot borrow as mutable
+}
+```
+
+The code still needed some change because the `future` in the  `Arc<Task>` became immutable. 
+
+"Okay. I can guarantee `Task` is created from a `Pin<Box<Future>>` and I think the same future will not be polled concurrently in two threads. So let me bypass the safety checks." Barbara confidently used some `unsafe` blocks to make it compile.
+
+```rust
+pub struct Task {
+    future: *mut (dyn Future<Output = ()> + Send + 'static),
+    ...
+}
+
+unsafe impl Send for Task {}
+unsafe impl Sync for Task {}
+
+pub fn poll_task(task: Arc<Task>) {
+    ...
+    unsafe {
+        Pin::new_unchecked(&mut *task.future).poll(&mut cx);
+    }
+}
+```
+
+Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of it and a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.
+
+Barbara soon got a idea to solve it. She added a state field to the `Task`. By carefully maintaining the state of the task, she could guarantee there are no duplicate tasks in the scheduler and no race can happen when polling the future:
+
+```rust
+const NOTIFIED: u64 = 1;
+const IDLE: u64 = 2;
+const POLLING: u64 = 3;
+const COMPLETED: u64 = 4;
+
+pub struct Task {
+    ...
+    state: AtomicU64,
+}
+
+impl Wake for Task {
+    fn wake(self: Arc<Self>) {
+        let mut state = self.state.load(Relaxed);
+        loop {
+            match state {
+                IDLE => match self
+                    .state
+                    .compare_exchange_weak(IDLE, NOTIFIED, AcqRel, Acquire)
+                {
+                    Ok(_) => self.scheduler.send(self.clone()),
+                    Err(s) => state = s,
+                },
+                POLLING => match self
+                    .state
+                    .compare_exchange_weak(POLLING, NOTIFIED, AcqRel, Acquire)
+                {
+                    Ok(_) => break,
+                    Err(s) => state = s,
+                },
+                _ => break,
+            }
+        }
+    }
+}
+pub fn poll_task(task: Arc<Task>) {
+    let waker = Waker::from(task.clone());
+    let mut cx = Context::from_waker(&waker);
+    loop {
+        task.state.store(POLLING, Release);
```

Comment:
I got tripped up wondering why you didn't need to read before polling (it's because wake() prevents the task from appearing in the scheduler twice). You do state this in the prose above, but maybe a comment here would help as well.